File: python/app/models/task.py
"""
Export task model for database
"""
from datetime import datetime
from enum import Enum
from typing import Optional

from tortoise import fields
from tortoise.models import Model


class TaskStatus(str, Enum):
    """Task status enumeration"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class ExportTask(Model):
    """
    Export task model for tracking async export operations
    """
    # Primary key
    id = fields.BigIntField(pk=True, description="任务ID")
    
    # Task identification
    task_id = fields.CharField(max_length=64, unique=True, index=True, description="任务UUID")
    download_token = fields.CharField(max_length=64, null=True, unique=True, description="下载令牌")
    
    # Task info
    task_type = fields.CharField(max_length=32, default="export", description="任务类型")
    export_format = fields.CharField(max_length=10, default="csv", description="导出格式 csv/xlsx")
    
    # Status
    status = fields.CharEnumField(TaskStatus, default=TaskStatus.PENDING, description="任务状态")
    progress = fields.IntField(default=0, description="进度百分比 0-100")
    total_count = fields.IntField(default=0, description="总记录数")
    processed_count = fields.IntField(default=0, description="已处理记录数")
    
    # File info
    file_path = fields.CharField(max_length=512, null=True, description="文件路径")
    file_size = fields.BigIntField(null=True, description="文件大小(字节)")
    
    # Error info
    error_message = fields.TextField(null=True, description="错误信息")
    
    # Query parameters (JSON)
    query_params = fields.JSONField(null=True, description="查询参数")
    
    # API Key
    api_key = fields.CharField(max_length=128, null=True, description="API Key")
    
    # Timestamps
    created_at = fields.DatetimeField(auto_now_add=True, description="创建时间")
    updated_at = fields.DatetimeField(auto_now=True, description="更新时间")
    started_at = fields.DatetimeField(null=True, description="开始时间")
    completed_at = fields.DatetimeField(null=True, description="完成时间")
    
    # Expiration
    expires_at = fields.DatetimeField(null=True, description="过期时间")

    class Meta:
        table = "export_tasks"
        table_description = "导出任务表"
        indexes = [
            ("task_id",),
            ("download_token",),
            ("status",),
            ("created_at",),
        ]

    def __str__(self) -> str:
        return f"ExportTask({self.task_id})"

    def __repr__(self) -> str:
        return (
            f"<ExportTask id={self.id} task_id={self.task_id} "
            f"status={self.status} progress={self.progress}%>"
        )

    @property
    def is_completed(self) -> bool:
        """Check if task is completed"""
        return self.status == TaskStatus.COMPLETED

    @property
    def is_failed(self) -> bool:
        """Check if task is failed"""
        return self.status == TaskStatus.FAILED

    @property
    def is_pending(self) -> bool:
        """Check if task is pending"""
        return self.status == TaskStatus.PENDING

    @property
    def is_processing(self) -> bool:
        """Check if task is processing"""
        return self.status == TaskStatus.PROCESSING

    async def start(self) -> None:
        """Mark task as started"""
        self.status = TaskStatus.PROCESSING
        self.started_at = datetime.now()
        await self.save(update_fields=["status", "started_at", "updated_at"])

    async def complete(self, file_path: str, file_size: int) -> None:
        """Mark task as completed"""
        self.status = TaskStatus.COMPLETED
        self.progress = 100
        self.file_path = file_path
        self.file_size = file_size
        self.completed_at = datetime.now()
        await self.save(update_fields=[
            "status", "progress", "file_path", "file_size", 
            "completed_at", "updated_at"
        ])

    async def fail(self, error_message: str) -> None:
        """Mark task as failed"""
        self.status = TaskStatus.FAILED
        self.error_message = error_message
        self.completed_at = datetime.now()
        await self.save(update_fields=["status", "error_message", "completed_at", "updated_at"])

    async def update_progress(self, processed: int, total: int) -> None:
        """Update task progress"""
        self.processed_count = processed
        self.total_count = total
        if total > 0:
            self.progress = int((processed / total) * 100)
        await self.save(update_fields=["processed_count", "total_count", "progress", "updated_at"])

File: rust/src/models/task.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// 导出任务状态
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum TaskStatus {
    Pending,
    Processing,
    Completed,
    Failed,
}

impl std::fmt::Display for TaskStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            TaskStatus::Pending => write!(f, "pending"),
            TaskStatus::Processing => write!(f, "processing"),
            TaskStatus::Completed => write!(f, "completed"),
            TaskStatus::Failed => write!(f, "failed"),
        }
    }
}

/// 导出任务
#[derive(Debug, Clone, Serialize)]
pub struct ExportTask {
    pub task_id: Uuid,
    pub status: TaskStatus,
    pub progress: u32,
    pub total: u32,
    pub format: String,
    pub created_at: DateTime<Utc>,
    pub completed_at: Option<DateTime<Utc>>,
    pub file_path: Option<String>,
    pub download_token: Option<String>,
    pub error_message: Option<String>,
}

impl ExportTask {
    pub fn new(format: String) -> Self {
        Self {
            task_id: Uuid::new_v4(),
            status: TaskStatus::Pending,
            progress: 0,
            total: 0,
            format,
            created_at: Utc::now(),
            completed_at: None,
            file_path: None,
            download_token: None,
            error_message: None,
        }
    }

    pub fn progress_percent(&self) -> f32 {
        if self.total == 0 {
            0.0
        } else {
            (self.progress as f32 / self.total as f32) * 100.0
        }
    }
}

/// 导出请求参数
#[derive(Debug, Deserialize)]
pub struct ExportRequest {
    pub format: Option<String>,
    pub start_time: Option<String>,
    pub end_time: Option<String>,
    pub order_status: Option<String>,
    pub min_amount: Option<f64>,
    pub max_amount: Option<f64>,
    pub user_id: Option<i64>,
    pub order_no: Option<String>,
}

impl Default for ExportRequest {
    fn default() -> Self {
        Self {
            format: Some("csv".to_string()),
            start_time: None,
            end_time: None,
            order_status: None,
            min_amount: None,
            max_amount: None,
            user_id: None,
            order_no: None,
        }
    }
}

/// 任务状态响应
#[derive(Debug, Serialize)]
pub struct TaskStatusResponse {
    pub task_id: Uuid,
    pub status: TaskStatus,
    pub progress: u32,
    pub total: u32,
    pub progress_percent: f32,
    pub download_url: Option<String>,
    pub error_message: Option<String>,
}

impl From<ExportTask> for TaskStatusResponse {
    fn from(task: ExportTask) -> Self {
        Self {
            task_id: task.task_id,
            status: task.status.clone(),
            progress: task.progress,
            total: task.total,
            progress_percent: task.progress_percent(),
            download_url: task.download_token.map(|token| {
                format!("/api/v1/exports/download/{}", token)
            }),
            error_message: task.error_message,
        }
    }
}

/// SSE 事件数据
#[derive(Debug, Serialize)]
pub struct SSEEventData {
    pub task_id: Uuid,
    pub status: TaskStatus,
    pub progress: u32,
    pub total: u32,
    pub progress_percent: f32,
    pub message: String,
}
